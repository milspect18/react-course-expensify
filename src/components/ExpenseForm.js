import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";


export default class ExpenseForm extends React.Component {
    state = {
        description: (this.props.expense ? this.props.expense.description : ""),
        note: (this.props.expense ? this.props.expense.note : ""),
        amount: (this.props.expense ? (this.props.expense.amount / 100.0).toString() : ""),
        createdAt: (this.props.expense ? moment(this.props.expense.createdAt) : moment()),
        dateFocused: false,
        formError: ""
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;

        // Only update if the input is a number with at most
        // two digits after the decimal place.  Paste the regex
        // into regex101.com for depper explanation
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ dateFocused: focused }));
    }

    onFormSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ formError: "Please provide a description and amount." }));
        } else {
            this.setState(() => ({ formError: "" }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseInt(parseFloat(this.state.amount) * 100),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    render() {
        return (
            <form className="form" onSubmit={this.onFormSubmit}>
                {this.state.formError && <p className="form__error">{this.state.formError}</p>}
                <input
                    type="text"
                    className="text-input"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    className="text-input"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.dateFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    className="text-area"
                    placeholder="Add a note for your expense (optional"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>

                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        );
    }
}