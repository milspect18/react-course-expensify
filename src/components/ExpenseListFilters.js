import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters';
import { setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';


export class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null,
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onDatePickerFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSelectChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    }

    render() {
        return (
            <div>
                <input
                    type='text'
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSelectChange}
                >
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId={'expenseListFilters_startDateID'}
                    endDate={this.props.filters.endDate}
                    endDateId={'expenseListFilters_endDateID'}
                    onDatesChange={this.onDatesChange}
                    onFocusChange={this.onDatePickerFocusChange}
                    focusedInput={this.state.calenderFocused}
                    isOutsideRange={() => false}
                    numberOfMonths={1}
                    showClearDates={true}
                />
            </div>
        );
    };
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
}

const mapDispatchToProps = (dispatch) => {
    return {
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate)),
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount())
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
