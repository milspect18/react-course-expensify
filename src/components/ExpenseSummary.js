import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';


export const ExpenseSummary = ({ expenses }) => {
    const numExpenses = expenses.length;
    const total = selectExpensesTotal(expenses);

    return (
        <div>
            {`Viewing ${numExpenses} ${numExpenses === 1 ? 'expense' : 'expenses'} `}
            {`totalling ${numeral(total / 100).format('$0,0.00')}`}
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
}

export default connect(mapStateToProps)(ExpenseSummary);
