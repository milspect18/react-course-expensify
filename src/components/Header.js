import React from 'react';
import { Link } from 'react-router-dom';
import { startLogout } from '../actions/auth';
import { connect } from 'react-redux';

export const Header = ({ logout }) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to='/dashboard'>
                    <h1>Expensify</h1>
                </Link>
                <button
                    className="button button--link"
                    onClick={logout}
                >
                    Logout
                </button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    logout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);
