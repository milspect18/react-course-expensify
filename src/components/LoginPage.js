import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ login }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expensify</h1>
            <p>It is time to get your expenses under control!</p>
            <button
                className="button"
                onClick={login}
            >
                Login with Google
            </button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    login: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
