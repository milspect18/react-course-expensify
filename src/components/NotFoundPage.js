import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <div>
        That aint here man...
        <Link to='/' >Go Home</Link>
    </div>
);

export default NotFoundPage;
