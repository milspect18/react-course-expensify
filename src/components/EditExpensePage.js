import React from 'react';
import { connect } from 'react-redux';
import { editExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
    onSubmit = (updatedExpense) => {
        this.props.editExpense(this.props.expenseToEdit.id, updatedExpense);
        this.props.history.push('/');
    };

    onClick = () => {
        this.props.startRemoveExpense(this.props.expenseToEdit.id);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expenseToEdit}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick} >
                    Remove
                </button>
            </div>
        );
    }
}

const maptStateToProps = (state, props) => {
    return {
        expenseToEdit: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        editExpense: (id, newDetails) => dispatch(editExpense(id, newDetails)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense({ id }))
    };
};

export default connect(maptStateToProps, mapDispatchToProps)(EditExpensePage);
