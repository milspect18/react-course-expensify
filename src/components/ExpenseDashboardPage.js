import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseSummary from '../components/ExpenseSummary';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseSummary />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;
