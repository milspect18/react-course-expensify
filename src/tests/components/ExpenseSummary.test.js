import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';


test('should render with single expense (singular text)', () => {
    const wrapper = shallow(
        <ExpenseSummary
            expensesCount={1}
            expensesTotal={195}
        />
    );

    expect(wrapper).toMatchSnapshot();
});


test('should render with multiple expenses (plural text)', () => {
    const wrapper = shallow(
        <ExpenseSummary
            expensesCount={3}
            expensesTotal={226762}
        />
    );

    expect(wrapper).toMatchSnapshot();
});
