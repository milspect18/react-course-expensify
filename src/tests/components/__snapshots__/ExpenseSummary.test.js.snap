// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with multiple expenses (plural text) 1`] = `
<div>
  <h1>
    Viewing 3 expenses totalling $2,267.62
  </h1>
</div>
`;

exports[`should render with single expense (singular text) 1`] = `
<div>
  <h1>
    Viewing 1 expense totalling $1.95
  </h1>
</div>
`;
